### PR TITLE
Ban registering obligations during InferCtxt snapshots.

### PR DESCRIPTION
**Pull request by eddyb**
Back in #33852, a flag was added to `InferCtxt` to prevent rolling back a snapshot if obligations were added to some `FulfillmentContext` during the snapshot, to prevent leaking fresh inference variables (created during that snapshot, so their indices would get reused) in obligations, which could ICE or worse.

But that isn't enough in the long run, as type-checking ends up relying on success implying that eager side-effects are fine, and while stray obligations *do* get caught nowadays, those errors prevent, e.g. the speculative coercions from #37658, which *have to* be rolled back *even* if they succeed.

We can't just allow those obligations to stay around though, because we end up, again, in ICEs or worse.
Instead, this PR modifies `lookup_method_in_trait_adjusted` to return `InferOk` containing the obligations that `Autoderef::finalize_as_infer_ok` can propagate to deref coercions.

As there shouldn't be *anything* left that registers obligations during snapshots, it's completely banned.

r? @nikomatsakis @arielb1

### CODE REVIEW

File: src/librustc/infer/mod.rs [hunk 1]
```diff
@@ -861,46 +863,45 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
...
-    // useful, much better than grovelling through megabytes of
-    // RUST_LOG output.
+    // snapshot, create type variables, and register obligations
+    // which may involve those type variables in the fulfillment cx,
+    // potentially leaving "dangling type variables" behind.
+    // In such cases, an assertion will fail when attempting to
+    // register obligations, within a snapshot. Very useful, much
+    // better than grovelling through megabytes of RUST_LOG output.
     //
-    // HOWEVER, in some cases the flag is wrong. In particular, we
```
Comment (arielb1) [inline src/librustc/infer/mod.rs]: Maybe it's better to have a `within_snapshot` boolean on `FulfillmentCx`?

Comment (eddyb) [inline src/librustc/infer/mod.rs]: Yeah it's not super clear what the best option is. It's probably a good idea to tie fulfillment contexts to inference contexts (through lifetimes?) and maybe have a snapshot depth... or maybe an "epoch"?

**Comment by eddyb** [comment]
[Crater run](https://gist.github.com/eddyb/2329f42a26473c55c522e9cb9fd7aa20) has found [one root regression](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2FJ_rliK2QQ9WzD2QtKkWljw%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=96&highlightEnd=96&lineNumber=&wrapLines=false&showLineNumbers=true&jumpToHighlight=true&followLog=false&lineHeight=19&asText=false).

File: src/librustc/infer/mod.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc/infer/mod.rs]: Longer term, I have been hoping to just get rid of the notion of snapshots except for very short-term (e.g., to make a unification operation atomic) and then create nested fulfillment contexts for "probing sessions".

Comment (nikomatsakis) [inline src/librustc/infer/mod.rs]: Well, maybe that's not a complete answer. Still, I guess I hope that we can move things into nested contexts and generally remove the need for snapshots.

**Review by nikomatsakis — APPROVED** [review]
Looks good to me, do we want to do a crater run or anything like that? I could imagine some subtle effects (as evidenced by the last commit).

**Comment by nikomatsakis** [comment]
Ah, I see @eddyb is ahead of me here. Very good.

**Comment by nikomatsakis** [comment]
Actually, @arielb1, were you satisfied? (If so, feel free to r=me, or you :)

**Pull request merged**